src/Castle.ts

```diff
@@ -0,0 +1,198 @@
+import { IncomingHttpHeaders } from 'http2';
+
+const apiUrl = 'https://api.castle.com';
+
+type TrackParameters = {
+  event: string;
+  user_id: string;
+  user_traits: any;
+  created_at?: string;
+  context: {
+    ip: string;
+    client_id: string;
+    headers: IncomingHttpHeaders;
+  };
+};
+
+type AuthenticateResult = {
+  action: string;
+  user_id?: string;
+  device_token?: string;
+  failover?: boolean;
+  failover_reason?: string;
+};
+
+export class Castle {
+  private apiSecret: string;
+  private timeout: number;
+  private allowedHeaders: string[];
+  private disallowedHeaders: string[];
+  private overrideFetch: any;
+  private failoverStrategy: string;
+
+  constructor({
+    apiSecret,
+    timeout = 1000,
+    allowedHeaders = [],
+    disallowedHeaders = [],
+    overrideFetch = fetch,
+    failoverStrategy = 'allow',
+  }: {
+    apiSecret: string;
+    timeout?: number;
+    allowedHeaders?: string[];
+    disallowedHeaders?: string[];
+    overrideFetch?: any;
+    failoverStrategy?: string;
+  }) {
+    if (!apiSecret) {
+      throw new Error(
+        'Castle: Unable to instantiate Castle client, API secret is missing.'
+      );
+    }
+
+    this.apiSecret = apiSecret;
+    this.timeout = timeout;
+    this.allowedHeaders = allowedHeaders;
+    this.disallowedHeaders = disallowedHeaders;
+    this.overrideFetch = overrideFetch;
+    this.failoverStrategy = failoverStrategy;
+  }
+
+  public async authenticate(
+    params: TrackParameters
+  ): Promise<AuthenticateResult> {
+    if (!params.event) {
+      throw new Error('Castle: event is required when calling authenticate.');
+    }
+
+    let response: Response;
+
+    try {
+      response = await this.getFetch()(`${apiUrl}/v1/authenticate`, {
+        method: 'POST',
+        timeout: this.timeout,
+        headers: this.generateDefaultRequestHeaders(),
+        body: this.generateRequestBody(params),
+      });
+    } catch (e) {
+      if (e.message.startsWith('network timeout')) {
+        return {
+          action: this.failoverStrategy,
+          failover: true,
+          failover_reason: 'timeout',
+        };
+      } else {
+        throw e;
+      }
+    }
+
+    this.handleUnauthorized(response);
+
+    if (!response.ok) {
+      throw new Error(
+        `Castle: \`/authenticate\` request unsuccessful. Expected ok result, got ${
+          response.status
+        }`
+      );
+    }
+
+    return response.json();
+  }
+
+  public async track(params: TrackParameters) {
+    if (!params.event) {
+      throw new Error('Castle: event is required when calling track.');
+    }
+
+    let response: Response;
+
+    try {
+      response = await this.getFetch()(`${apiUrl}/v1/track`, {
+        method: 'POST',
+        timeout: this.timeout,
+        headers: this.generateDefaultRequestHeaders(),
+        body: this.generateRequestBody(params),
+      });
+    } catch (e) {
+      if (e.message.startsWith('network timeout')) {
```

Comment:
duplication - can we extract the error handling as it seems to be used multiple times?

Reply:
Could do `if (isNetworkTimeout(e)) {`. It's really minor, but sure, why not!

Reply:
It is minor but makes life easier :) thank you!